src/corecel/data/LdgIterator.hh

```diff
@@ -0,0 +1,253 @@
+//---------------------------------*-CUDA-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file corecel/data/LdgIterator.hh
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include <cstddef>
+#include <iterator>
+#include <type_traits>
+
+#include "corecel/Macros.hh"
+#include "corecel/Types.hh"
+#include "corecel/math/Algorithms.hh"
+
+namespace celeritas
+{
+namespace detail
+{
+//---------------------------------------------------------------------------//
+
+/*!
+ * Reads a value T using __ldg builtin and return a copy of it
+ */
+template<class T, typename = void>
+struct LDGLoad
+{
+    using value_type = T;
+    using pointer = std::add_pointer_t<value_type const>;
+    using reference = value_type;
+
+    CELER_CONSTEXPR_FUNCTION static reference read(pointer p)
+    {
+#if CELER_DEVICE_COMPILE
+        return __ldg(p);
+#else
+        return *p;
+#endif
+    }
+};
+
+/*!
+ * Specialization when T == OpaqueId.
+ * Wraps the underlying index in a OpaqueId when returning it.
+ */
+template<class T>
+struct LDGLoad<T, std::enable_if_t<is_opaqueid_v<T>>>
```

Comment:
Do you write it like this instead of
```c++
template<class I, class T>
struct LDGLoad<OpaqueId<I, T>, void>
```
because it's more similar to C++20 concepts? That's how I have `detail::RangeTypeTraits` set up. Speaking of, we'll probably want another specialization for enumerations... that can go in a follow-up commit of course.

Reply:
that was the idea but using specialization only without resorting to SFINAE is better.